File: app/channels/conversation_channel.rb
class ConversationChannel < ApplicationCable::Channel
  def subscribed
    stream_for Conversation.find(params[:conversation_id])
  end

  def typing(data)
    ConversationChannel.broadcast_to(
      Conversation.find(params[:conversation_id]),
      { typing: true, sender: current_user&.id || current_vendor&.id }
    )
  end

  def stop_typing(data)
    ConversationChannel.broadcast_to(
      Conversation.find(params[:conversation_id]),
      { typing: false, sender: current_user&.id || current_vendor&.id }
    )
  end
end

File: app/javascript/channels/conversation_channel.js
import consumer from "./consumer"

const conversationId = document.getElementById("conversation").dataset.id

const channel = consumer.subscriptions.create(
  { channel: "ConversationChannel", conversation_id: conversationId },
  {
    received(data) {
      if (data.typing) {
        document.getElementById("typing-indicator").innerText = "Typing..."
      } else {
        document.getElementById("typing-indicator").innerText = ""
      }
    }
  }
)

document.getElementById("message_body").addEventListener("input", () => {
  channel.perform("typing", {})
  clearTimeout(window.typingTimeout)
  window.typingTimeout = setTimeout(() => channel.perform("stop_typing", {}), 2000)
})
